src/collectors/uniswapx_route_collector.rs

```diff
@@ -166,33 +168,58 @@ impl UniswapXRouteCollector {
-// TODO: implement order deduplication
 #[async_trait]
 impl Collector<RoutedOrder> for UniswapXRouteCollector {
     async fn get_event_stream(&self) -> Result<CollectorStream<'_, RoutedOrder>> {
         let stream = async_stream::stream! {
-            let mut receiver = self.route_request_receiver.lock().await;
-            while let Some(route_requests) = receiver.recv().await {
-                let tasks: FuturesUnordered<_> = route_requests.into_iter()
+            loop {
+                let mut all_requests = Vec::new();
+                let mut receiver = self.route_request_receiver.lock().await;
+
+                // Collect all available messages without blocking
+                while let Ok(requests) = receiver.try_recv() {
+                    all_requests.extend(requests);
+                }
+
+                // If no messages were received, wait for one
+                if all_requests.is_empty() {
+                    if let Some(requests) = receiver.recv().await {
+                        all_requests.extend(requests);
+                    } else {
+                        break; // Channel closed
+                    }
+                }
+
+                drop(receiver); // Release the lock
+
+                // deduplicate requests on first order's order_hash
+                let mut seen = HashSet::new();
+                all_requests = all_requests.into_iter().filter(|req| seen.insert(req.orders[0].hash.clone())).collect();
+
+                // Process all_requests
+                let tasks: Vec<_> = all_requests.into_iter()
                     .map(|batch| {
                         let OrderBatchData { orders, token_in, token_out, amount_in, .. } = batch.clone();
                         info!(
-                            "Checking batch: {} orders, token in: {}, token out: {}",
-                            orders.len(), token_in, token_out
+                            "{} - Routing order, token in: {}, token out: {}",
+                            orders[0].hash,
+                            token_in, token_out
                         );
 
                         async move {
-                            (batch, route_order(RouteOrderParams {
+                            let route_result = route_order(RouteOrderParams {
                                 chain_id: self.chain_id,
                                 token_in: token_in.clone(),
                                 token_out: token_out.clone(),
                                 amount: amount_in.to_string(),
                                 recipient: self.executor_address.clone(),
-                            }).await)
+                            }).await;
+                            (batch, route_result)
                         }
-                    }).collect();
+                    })
+                    .collect();
 
-                let routes: Vec<_> = tasks.collect().await;
+                let routes = join_all(tasks).await;
```

Comment:
I wonder how hard it is to yield RoutedOrder as they become available, instead of waiting for all futures to complete

Reply:
yea good suggestion!